src/java.base/share/classes/java/lang/invoke/BootstrapMethodInvoker.java

```diff
@@ -133,40 +140,40 @@ else if (info.getClass() == int[].class) {
                 } else if (isStringConcatFactoryBSM(bsmType) && argv.length >= 1) {
                     String recipe = (String)argv[0];
                     Object[] shiftedArgs = Arrays.copyOfRange(argv, 1, argv.length);
+                    maybeReBoxElements(shiftedArgs);
                     result = (CallSite)bootstrapMethod.invokeExact(caller, name, (MethodType)type, recipe, shiftedArgs);
                 } else if (isLambdaMetafactoryAltMetafactoryBSM(bsmType)) {
+                    maybeReBoxElements(argv);
                     result = (CallSite)bootstrapMethod.invokeExact(caller, name, (MethodType)type, argv);
                 } else {
-                    switch (argv.length) {
-                        case 0:
-                            result = invoke(bootstrapMethod, caller, name, type);
-                            break;
-                        case 1:
-                            result = invoke(bootstrapMethod, caller, name, type,
-                                            argv[0]);
-                            break;
-                        case 2:
-                            result = invoke(bootstrapMethod, caller, name, type,
-                                            argv[0], argv[1]);
-                            break;
-                        case 3:
-                            result = invoke(bootstrapMethod, caller, name, type,
-                                            argv[0], argv[1], argv[2]);
-                            break;
-                        case 4:
-                            result = invoke(bootstrapMethod, caller, name, type,
-                                            argv[0], argv[1], argv[2], argv[3]);
-                            break;
-                        case 5:
-                            result = invoke(bootstrapMethod, caller, name, type,
-                                            argv[0], argv[1], argv[2], argv[3], argv[4]);
-                            break;
-                        case 6:
-                            result = invoke(bootstrapMethod, caller, name, type,
-                                            argv[0], argv[1], argv[2], argv[3], argv[4], argv[5]);
-                            break;
-                        default:
-                            result = invokeWithManyArguments(bootstrapMethod, caller, name, type, argv);
+                    maybeReBoxElements(argv);
+                    if (argv.length > 6) {
+                        result = invokeWithManyArguments(bootstrapMethod, caller, name, type, argv);
```

Comment:
it'd be cleaner to move this to the default case in line 162 and 174 instead of having this special if-block.

Reply:
Good catch!